src/analysis/processing/qgsalgorithmdbscanclustering.cpp

```diff
@@ -0,0 +1,305 @@
+}
+
+QString QgsDbscanClusteringAlgorithm::group() const
+{
+  return QObject::tr( "Vector analysis" );
+}
+
+QString QgsDbscanClusteringAlgorithm::groupId() const
+{
+  return QStringLiteral( "vectoranalysis" );
+}
+
+void QgsDbscanClusteringAlgorithm::initAlgorithm( const QVariantMap & )
+{
+  addParameter( new QgsProcessingParameterFeatureSource( QStringLiteral( "INPUT" ),
+                QObject::tr( "Input layer" ), QList< int >() << QgsProcessing::TypeVectorPoint ) );
+  addParameter( new QgsProcessingParameterNumber( QStringLiteral( "MIN_SIZE" ), QObject::tr( "Minimum cluster size" ),
+                QgsProcessingParameterNumber::Integer, 5, false, 1 ) );
+  addParameter( new QgsProcessingParameterDistance( QStringLiteral( "EPS" ),
+                QObject::tr( "Maximum distance between clustered points" ), 1, QStringLiteral( "INPUT" ), false, 0 ) );
+
+  auto dbscanStarParam = qgis::make_unique<QgsProcessingParameterBoolean>( QStringLiteral( "DBSCAN*" ),
+                         QObject::tr( "Treat border points as noise (DBSCAN*)" ), false, true );
+  dbscanStarParam->setFlags( dbscanStarParam->flags() | QgsProcessingParameterDefinition::FlagAdvanced );
+  addParameter( dbscanStarParam.release() );
+
+  auto fieldNameParam = qgis::make_unique<QgsProcessingParameterString>( QStringLiteral( "FIELD_NAME" ),
+                        QObject::tr( "Cluster field name" ), QStringLiteral( "CLUSTER_ID" ) );
+  fieldNameParam->setFlags( fieldNameParam->flags() | QgsProcessingParameterDefinition::FlagAdvanced );
+  addParameter( fieldNameParam.release() );
+
+  addParameter( new QgsProcessingParameterFeatureSink( QStringLiteral( "OUTPUT" ), QObject::tr( "Clusters" ), QgsProcessing::TypeVectorPoint ) );
+
+  addOutput( new QgsProcessingOutputNumber( QStringLiteral( "NUM_CLUSTERS" ), QObject::tr( "Number of clusters" ) ) );
+}
+
+QString QgsDbscanClusteringAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "Clusters point features based on a 2D implementation of Density-based spatial clustering of applications with noise (DBSCAN) algorithm.\n\n"
+                      "The algorithm requires two parameters, a minimum cluster size (“minPts”), and the maximum distance allowed between clustered points (“eps”)." );
+}
+
+QgsDbscanClusteringAlgorithm *QgsDbscanClusteringAlgorithm::createInstance() const
+{
+  return new QgsDbscanClusteringAlgorithm();
+}
+
+struct KDBushDataEqualById
+{
+  bool operator()( const QgsSpatialIndexKDBushData &a, const QgsSpatialIndexKDBushData &b ) const
+  {
+    return  a.id == b.id;
+  }
+};
+
+struct KDBushDataHashById
+{
+  std::size_t operator()( const QgsSpatialIndexKDBushData &a ) const
+  {
+    return std::hash< QgsFeatureId > {}( a.id );
+  }
+};
+
+bool operator <( const QgsSpatialIndexKDBushData &data, const QgsFeatureId id )
+{
+  return data.id < id;
+};
+
+bool operator <( const QgsFeatureId id, const QgsSpatialIndexKDBushData &data )
+{
+  return id < data.id;
+};
```

Comment:
Are the less than operators used anywhere? I guess they are not needed for unordered_set?

Reply:
I'll double check again. These may have only been required for an earlier iteration of this work.

Reply:
Thanks - this identified an issue in the code (was using std::set_difference, which shouldn't be used for unordered sets). Shaved about 25% off the processing time by fixing this :smile:

Reply:
Awesome!